### PR TITLE
Tooltip for actions

### PR DESCRIPTION
**Pull request by francocorreasosa**
You can use a special variation of the tooltip which is made for input actions.

It does recieve a `content` object with the following keys:

- `default`: Is the default text to show in the tooltip (same as `content` in a normal tooltip).
- `loading`: Is the text will show if the child button `onClick` handler is an async function (Promise) while it's loading.
- `success`: Is the text will show when the `onClick` returns (if it's a sync function) or resolves (if it's async).
- `error`: If the text will show if the `onClick` is an async function and is rejected.

#### Example

```jsx
<TextInput
  type="text"
  placeholder="Enter some text"
  actions={[
    <Tooltip.Action
      content={{
        default: 'Copy',
        success: 'Copied!'
      }}
    >
      <Button icon="copy" onClick={e => console.log(e)} />
    </Tooltip.Action>,
    <Tooltip.Action
      content={{
        default: 'Do work',
        loading: 'Working...',
        success: 'Done!',
        error: 'Could not work'
      }}
    >
      <Button icon="wrench" onClick={sampleAsyncAction} />
    </Tooltip.Action>,
    <Tooltip.Action
      content={{
        default: 'Do work',
        loading: 'Working...',
        success: 'Done!',
        error: 'Could not work'
      }}
    >
      <Button icon="reload" onClick={sampleAsyncFailingAction} />
    </Tooltip.Action>,
    <Button icon="delete" onClick={e => console.log(e)} />
  ]}
/>
```

demo

### CODE REVIEW

File: core/components/atoms/tooltip/action.js [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    return React.cloneElement(button, { onClick: newHandler, appearance: 'link' })
+  }
+
+  render() {
+    const { tooltipContent } = this.state
+    const { content, actionSuccess } = this.props
+    const children = this.processAction()
+
+    return (
+      <Tooltip content={tooltipContent ? tooltipContent : content.default} children={children} />
```
Comment (cristiandouce) [inline core/components/atoms/tooltip/action.js]: if `props.content` can be a string, then `content.default` could end up returning undefined, right?

File: core/components/atoms/tooltip/action.js [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+        })
+        .catch(() => this.setTooltipContent(content.error || 'Something did not work'))
+    }
+
+    return React.cloneElement(button, { onClick: newHandler, appearance: 'link' })
+  }
+
+  render() {
+    const { tooltipContent } = this.state
+    const { content, actionSuccess } = this.props
```
Comment (cristiandouce) [inline core/components/atoms/tooltip/action.js]: `actionSuccess` is not defined on `PropTypes`.
also, `content` should be `this.processContent()`, right?
```suggestion
    const content = this.processContent()
```

File: core/components/atoms/tooltip/tooltip.js [hunk 3]
```diff
@@ -94,6 +95,15 @@ class Tooltip extends React.Component {
   }
 }
 
+Tooltip.ObservedElement = class extends React.Component {
```
Comment (cristiandouce) [inline core/components/atoms/tooltip/tooltip.js]: 💇‍♂️

File: core/components/atoms/tooltip/tooltip.md [hunk 4]
```diff
@@ -44,3 +44,54 @@ The Tooltip wraps long text by default to make it readable. Try to keep the cont
...
+    </Tooltip.Action>,
+    <Tooltip.Action
+      content={{
+        default: 'Do work',
+        loading: 'Working...',
+        success: 'Done!',
+        error: 'Could not work'
+      }}
+    >
+      <Button icon="reload" onClick={sampleAsyncFailingAction} />
```
Comment (cristiandouce) [inline core/components/atoms/tooltip/tooltip.md]: could we have an example where the button has a disabled/loading state as well?

File: core/components/atoms/tooltip/action.js (hunk 1 above)
Comment (francocorreasosa) [reply core/components/atoms/tooltip/action.js]: Good catch ✅

File: core/components/atoms/tooltip/action.js [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+    const content = this.preprocessContent()
+
+    const newHandler = event => {
+      if (content.loading) {
+        this.setTooltipContent(content.loading, { resetDelay: loadingTimeout || 5000 })
+      }
+
+      Promise.resolve(button.props.onClick(event))
+        .then(result => {
+          this.setTooltipContent(content.success)
```
Comment (cristiandouce) [inline core/components/atoms/tooltip/action.js]: what if this is undefined?
I'd probably avoid this, or just use `content.default` as fallback

File: core/components/atoms/tooltip/action.js [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+  /**
+   * Sets the tooltip inner text to `newContent`
+   * and resets the content to `props.content.default` after `resetDelay`.
+   * @param {string} newContent - Text to be set in the tooltip
+   * @param {string} resetDelay - Time in milliseconds to reset back to default text.
+   */
+  setTooltipContent(newContent, { resetDelay = 3000 } = {}) {
+    const content = this.preprocessContent()
+    this.setState({ tooltipContent: newContent })
+    setTimeout(() => this.setState({ tooltipContent: content.default }), resetDelay)
```
Comment (cristiandouce) [inline core/components/atoms/tooltip/action.js]: I'd save this timeout id somewhere if possible, and on `componentWillUnmount` just reset everything?

Comment (cristiandouce) [reply core/components/atoms/tooltip/action.js]: should we handle something on `onMouseLeave` or whatever trigger we have for the tooltip, cancel this timeout?

File: core/components/atoms/tooltip/action.js (hunk 5 above)
Comment (francocorreasosa) [reply core/components/atoms/tooltip/action.js]: Done ✅

File: core/components/atoms/tooltip/tooltip.md (hunk 4 above)
Comment (francocorreasosa) [reply core/components/atoms/tooltip/tooltip.md]: Done ✅

**Comment by francocorreasosa** [comment]
@landitus hmm 🤔 can't reproduce here. happy to pair with you tomorrow so we can take a look in your computer together.

kapture 2019-02-21 at 20 40 15

**Comment by siddharthkp** [comment]
👋 API suggestion:

`Tooltip.Action` sounds like it renders the button too

Should we call it `Tooltip.ActionWrapper` which conveys that you use it with Actions?

**Comment by francocorreasosa** [comment]
@siddharthkp:

R1. That's fair enough, let's do it 👍 

R2. This solution is not meant to be used with switches, just buttons. I haven't seen any example of an input using a switch inside actually 🤔

File: core/components/atoms/tooltip/action.js [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+    this.setState({ tooltipContent: newContent })
+
+    this.clearTimer()
+
+    if (newContent === content.default) {
+      return
+    }
+
+    this.timer = setTimeout(
+      () => this.setState({ tooltipContent: content.default, ...additionalState }),
```
Comment (cristiandouce) [inline core/components/atoms/tooltip/action.js]:
```suggestion
      () => this.setState({ tooltipContent: content.default }),
```

File: core/components/atoms/tooltip/action.js [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+  /**
+   * Sets the tooltip inner text to `newContent`
+   * and resets the content to `props.content.default` after `resetDelay`.
+   * @param {string} newContent - Text to be set in the tooltip
+   * @param {string} resetDelay - Time in milliseconds to reset back to default text.
+   * @param {object} additionalState - Any additional data you want to set in the state.
+   */
+  setTooltipContent(newContent, { resetDelay = 3000 } = {}, additionalState = {}) {
+    const content = this.preprocessContent()
+    this.setState({ tooltipContent: newContent })
```
Comment (cristiandouce) [inline core/components/atoms/tooltip/action.js]:
```suggestion
    this.setState({ tooltipContent: newContent, ...additionalState })
```

File: core/components/atoms/tooltip/action.js [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+  }
+
+  /**
+   * Sets the tooltip inner text to `newContent`
+   * and resets the content to `props.content.default` after `resetDelay`.
+   * @param {string} newContent - Text to be set in the tooltip
+   * @param {string} resetDelay - Time in milliseconds to reset back to default text.
+   * @param {object} additionalState - Any additional data you want to set in the state.
+   */
+  setTooltipContent(newContent, { resetDelay = 3000 } = {}, additionalState = {}) {
```
Comment (cristiandouce) [inline core/components/atoms/tooltip/action.js]:
```suggestion
  setTooltipContent(newContent, additionalState = {}) {
```

File: core/components/atoms/tooltip/action.js [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+  preprocessContent() {
+    const rawContent = this.props.content
+    return typeof rawContent === 'string' ? { default: rawContent } : rawContent
+  }
+
+  /**
+   * Sets the tooltip inner text to `newContent`
+   * and resets the content to `props.content.default` after `resetDelay`.
+   * @param {string} newContent - Text to be set in the tooltip
+   * @param {string} resetDelay - Time in milliseconds to reset back to default text.
```
Comment (cristiandouce) [inline core/components/atoms/tooltip/action.js]:
```suggestion
```

File: core/components/atoms/tooltip/action.js [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+    const content = this.preprocessContent()
+    this.setState({ tooltipContent: newContent })
+
+    this.clearTimer()
+
+    if (newContent === content.default) {
+      return
+    }
+
+    this.timer = setTimeout(
```
Comment (cristiandouce) [inline core/components/atoms/tooltip/action.js]:
```suggestion
    const { resetDelay } = this.props

    this.timer = setTimeout(
```

File: core/components/atoms/tooltip/action.js [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+   * from a text input / text area action.
+   */
+  processAction() {
+    const { children: button, resetDelay } = this.props
+    const content = this.preprocessContent()
+
+    const newHandler = event => {
+      this.resetState()
+      if (content.loading) {
+        this.setTooltipContent(content.loading, { resetDelay: resetDelay }, { taskIsRunning: true })
```
Comment (cristiandouce) [inline core/components/atoms/tooltip/action.js]:
```suggestion
        this.setTooltipContent(content.loading, { taskIsRunning: true })
```

**Comment by cristiandouce** [comment]
> 👋 API suggestion:
> 
> Q1. Should we call it `Tooltip.ActionWrapper` which conveys that you use it with actions? `Tooltip.Action` sounds like it renders the button too. The idea is to communicate you bring your own actions.
> 
> Q2: Does this assume/enforce the child is a `Button` with `onClick` or are we open to accepting other components like `Switch` with `onChange`?

Q1: We should bring this up with the crew. I personally like `Tootip.Actions`.

Q2: We could discuss expanding the support in the future. Now it's just a very specific implementation to unblock us.

**Pull request merged**